File: CoreAnimationAdvancedTechniques/03_StrechableButtons/StretchableButtons/StretchableButtons/ViewController.m
//
//  ViewController.m
//  StretchableButtons
//
//  Created by ShannonChen on 2016/11/27.
//  Copyright © 2016年 ShannonChen. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()
@property (weak, nonatomic) IBOutlet UIButton *button1;
@property (weak, nonatomic) IBOutlet UIButton *button2;

@end

@implementation ViewController

- (void)addStretchableImage:(UIImage *)image
          withContentCenter:(CGRect)rect
                    toLayer:(CALayer *)layer {
    //set image
    layer.contents = (__bridge id)image.CGImage;
    
    //set contentsCenter
    layer.contentsCenter = rect;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    //load button image
    UIImage *image = [UIImage imageNamed:@"Button.png"];
    
    //set button 1
    [self addStretchableImage:image
            withContentCenter:CGRectMake(0.25, 0.25, 0.5, 0.5)
                      toLayer:self.button1.layer];
    
    //set button 2
    [self addStretchableImage:image
            withContentCenter:CGRectMake(0.25, 0.25, 0.5, 0.5)
                      toLayer:self.button2.layer];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: NullabilityDemo/NullabilityDemo/ViewController.swift
//
//  ViewController.swift
//  NullabilityDemo
//
//  Created by ShannonChen on 17/4/17.
//  Copyright © 2017年 ShannonChen. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        let aBook: Book = Book.init(name: "Bible")  // init!(name: String!)
        print(aBook.name);  //  var name: String! { get }
        
        let aBookWithAnnotations = BookWithAnnotations.init(name: "Notebook")  // init?(name: String)
        if let book = aBookWithAnnotations {
            print(book.name)  // var name: String { get }
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

